pkg/controller/kubedirectorcluster/cluster.go

```diff
@@ -57,6 +57,65 @@ func (r *ReconcileKubeDirectorCluster) syncCluster(
 		cr.Status = &kdv1.KubeDirectorClusterStatus{}
 		cr.Status.Roles = make([]kdv1.RoleStatus, 0)
 	}
+
+	// Check, if some statefulset is in the middle of upgrade/rollback process.
+	// If so, the current cluster state still can be in `configured` state, change it to `updating`
+	upgradeIsActive := false
+	for _, rs := range cr.Status.Roles {
+		if rs.RoleUpgradeStatus == kdv1.RoleRollingBack || rs.RoleUpgradeStatus == kdv1.RoleUpgrading {
+			upgradeIsActive = true
+			cr.Status.State = string(clusterUpdating)
```

Comment:
Hmm. I hadn't really thought about changing the top-level cluster state, but it does make sense. I'm curious though why it's necessary to flip it to "updating" here since it should already have been changed to "updating" when the upgradeInfo was created. That may be a sign that something is wrong about the flow through this overall syncClusters function.

Normally at the bottom of syncClusters, if the current cluster state is NOT "configured" then the code looks to see if syncClusterRoles has declared the need to create/delete any members (by the state value it returns). If not, then the logic is OK, our updating must be finished, so let's set the cluster state back to "configured". I think that may be what is happening to you.

(This would all be a little bit easier to see if we factor our the connections-related code there into a subroutine.)

I think we need to continue to restrict our cluster-state-change logic to that code near the end of syncCluster, and just update that logic.

The current code there looks like this:
```go
	// This "state" is from syncClusterRoles.
	if state == clusterMembersStableReady {
		if cr.Status.State != string(clusterReady) {
			// logging and a bunch of connections stuff
			cr.Status.State = string(clusterReady)
		}
		if currentHash == cr.Status.LastConnectionHash {
			// e.g. if we're not updating connections, leave the cluster as
			// "configured" and we're done.
			return nil
		}
	}

	// We reach this spot if syncClusterRoles determines that members are
	// changing, or if the connections changed (detected by that hash change).
	// In that case, unless we're currently in "creating" state we switch to
	// "updating"
	if cr.Status.State != string(clusterCreating) {
		cr.Status.State = string(clusterUpdating)
	}
```

So maybe it should look something like this:
```go
	upgradeOrRollbackActive := (cr.Status.UpgradeInfo != nil)

	if (state == clusterMembersStableReady) && !upgradeOrRollbackActive {
		if cr.Status.State != string(clusterReady) {
			// logging and a bunch of connections stuff
			cr.Status.State = string(clusterReady)
		}
		if currentHash == cr.Status.LastConnectionHash {
			return nil
		}
	}

	if cr.Status.State != string(clusterCreating) {
		cr.Status.State = string(clusterUpdating)
	}
```

If we handle the cluster state management down there -- which should manage both setting it to "updating" for an upgrade/rollback as well as clearing it back to "configured" when done -- then we shouldn't need to handle it anywhere in this chunk logic.

Reply:
Done